src/thrift-client/TgtInterfaceImpl.cpp

```diff
@@ -781,31 +737,71 @@ int64_t StordVmdk::RpcRequestScheduledCount() const noexcept {
 	return stats_.rpc_requests_scheduled_;
 }
 
-std::pair<Request*, bool> StordVmdk::NewRequest(Request::Type type,
-		const void* privatep, char* bufferp, size_t buf_sz, size_t xfer_sz,
-		int64_t offset) {
-	auto request = std::make_unique<Request>();
-	if (hyc_unlikely(not request)) {
-		return std::make_pair(nullptr, false);
+/* Prepare a sync request for completion */
+bool StordVmdk::PrepareRequest(std::shared_ptr<SyncRequest> request) {
+	bool prepared = true;
+	auto nreqp = request.get();
+
+	if (nreqp->type != Request::Type::kSync) {
+		return prepared;
 	}
 
-	++stats_.pending_;
+	std::lock_guard<std::mutex> lock(requests_.mutex_);
+	SyncRequest *reqp = dynamic_cast<SyncRequest*>(nreqp);
+	for (auto& req_map : requests_.scheduled_) {
+		auto req_ptr = req_map.second.get();
+		if ((req_ptr->type == Request::Type::kWrite or
+			req_ptr->type == Request::Type::kWriteSame) &&
+			req_ptr->IsOverlapped(reqp->offset, reqp->length)) {
+				req_ptr->reqp = request;
+				nreqp->count++;
+				/* Do not complete a sync request as overlapping writes on
+				   sync are not completed */
+				prepared = false;
+		}
+	}
+	requests_.sync_pending_.emplace(nreqp->id, std::move(request));
 
-	request->id = ++requestid_;
-	request->type = type;
-	request->privatep = privatep;
-	request->bufferp = bufferp;
-	request->buf_sz = buf_sz;
-	request->xfer_sz = xfer_sz;
-	request->offset = offset;
+	return prepared;
+}
+
+/* Prepare a request for RPC */
+bool StordVmdk::PrepareRequest(std::shared_ptr<Request> request) {
+	bool prepared = true;
+	auto nreqp = request.get();
+
+	++stats_.pending_;
 
-	auto reqp = request.get();
 	std::lock_guard<std::mutex> lock(requests_.mutex_);
-	bool schedule_now = requests_.scheduled_.empty() or
-		latency_avg_.Average() > kExpectedWanLatency;
+
+	/* Effective for new writes overlapping on sync */
+	bool overlapped_write = false;
+	if (nreqp->type == Request::Type::kWrite or
+		nreqp->type == Request::Type::kWriteSame) {
+		for (auto& sync_req : requests_.sync_pending_) {
+			SyncRequest *syncp =
+				dynamic_cast<SyncRequest *>(sync_req.second.get());
+			if (nreqp->IsOverlapped(syncp->offset, syncp->length)) {
```

Comment:
Incorrect use of dynamic_cast. It is not required here.

Reply:
Will remove it.